kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -42,9 +39,322 @@ import cats.syntax.either._
+                  s"Retrying all messages that could not be written to SQS queue ${sqs.sqsBufferName} in $nextBackoff milliseconds..."
+                else
+                  s"Sending ${s.size} messages that could not be written to SQS queue ${sqs.sqsBufferName} to Kinesis stream $streamName in $nextBackoff milliseconds..."
+              }
+              log.error(errorMessage)
+              scheduleWrite(s.map(_._1), target, nextBackoff)(retriesLeft - 1)
+            }
+          case Failure(f) =>
+            log.error("Writing failed with error:\n", f)
+            val errorMessage = {
+              if (retriesLeft > 1)
+                s"Retrying writing batch to SQS queue ${sqs.sqsBufferName} in $nextBackoff milliseconds..."
+              else
+                s"Sending batch to Kinesis stream $streamName in $nextBackoff milliseconds..."
+            }
+            log.error(errorMessage)
+            scheduleWrite(batch, target, nextBackoff)(retriesLeft - 1)
+        }
+    }
+
+  def writeBatchToKinesis(batch: List[Events]): Future[PutRecordsResult] =
+    Future {
+      val putRecordsRequest = {
+        val prr = new PutRecordsRequest()
+        prr.setStreamName(streamName)
+        val putRecordsRequestEntryList = batch.map { event =>
+          val prre = new PutRecordsRequestEntry()
+          prre.setPartitionKey(event.key)
+          prre.setData(ByteBuffer.wrap(event.payloads))
+          prre
+        }
+        prr.setRecords(putRecordsRequestEntryList.asJava)
+        prr
+      }
+      client.putRecords(putRecordsRequest)
+    }
+
+  def writeBatchToSqs(batch: List[Events], sqs: SqsClientAndName): Future[List[(Events, BatchResultErrorInfo)]] =
+    Future {
+      val splitBatch = split(batch, getByteSize, MaxSqsBatchSizeN, MaxSqsBatchSizeBytes)
+      splitBatch.map(toSqsMessages).flatMap { msgGroup =>
+        val entries = msgGroup.map(_._2)
+        val batchRequest =
+          new SendMessageBatchRequest().withQueueUrl(sqs.sqsBufferName).withEntries(entries.asJava)
+        val response = sqs.sqsClient.sendMessageBatch(batchRequest)
+        val failures = response
+          .getFailed
+          .asScala
+          .toList
+          .map { bree =>
+            (bree.getId, BatchResultErrorInfo(bree.getCode, bree.getMessage))
+          }
+          .toMap
+        // Events to retry and reasons for failure
+        msgGroup.collect {
+          case (e, m) if failures.contains(m.getId) =>
+            (e, failures(m.getId))
+        }
+      }
+    }
+
+  def toSqsMessages(events: List[Events]): List[(Events, SendMessageBatchRequestEntry)] =
+    events.map(e => (e, new SendMessageBatchRequestEntry(UUID.randomUUID.toString, b64Encode(e.payloads))))
+
+  def b64Encode(msg: Array[Byte]): String = {
+    val buffer = java.util.Base64.getEncoder.encode(msg)
+    new String(buffer)
+  }
+
+  def scheduleWrite(batch: List[Events], target: Target, lastBackoff: Long = minBackoff)(
+    retriesLeft: Int
```

Comment:
Why separating this parameter from the other ones ?

Reply:
I feel like it belongs as a last parameter, but we can't vary it's value without foregoing the default value of `lastBackoff` if we put them in the same param list.